mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp

```diff
@@ -0,0 +1,220 @@
+        // with the right shape and an identity layout map.
+        auto shape = memrefType.getShape();
+        // FlatAffineConstraint may later on use symbolicOperands.
+        FlatAffineConstraints fac(rank, 0);
+        for (unsigned d = 0; d < rank; ++d) {
+          fac.addConstantLowerBound(d, 0);
+          fac.addConstantUpperBound(d, shape[d] - 1);
+        }
+
+        // We compose this map with the original index (logical) space to derive
+        // the upper bounds for the new index space.
+        unsigned newRank = layoutMap.getNumResults();
+        if (failed(fac.composeMatchingMap(layoutMap)))
+          continue;
+        // TODO: semi-affine maps.
+
+        // Project out the old data dimensions.
+        fac.projectOut(newRank,
+                       fac.getNumIds() - newRank - fac.getNumLocalIds());
+        SmallVector<int64_t, 4> newShape(newRank);
+        for (unsigned d = 0; d < newRank; ++d) {
+          // The lower bound for the shape is always zero.
+          auto ubConst = fac.getConstantUpperBound(d);
+          // For a static memref and an affine map with no symbols, this is
+          // always bounded.
+          assert(ubConst.hasValue() && "should always have an upper bound");
+          if (ubConst.getValue() < 0)
+            // This is due to an invalid map that maps to a negative space.
+            continue;
+          newShape[d] = ubConst.getValue() + 1;
+        }
+
+        // Create the new memref type after trivializing the old layout map.
+        MemRefType newMemRefType =
+            MemRefType::Builder(memrefType)
+                .setShape(newShape)
+                .setAffineMaps(b.getMultiDimIdentityMap(newRank));
+
+        auto oldMemRef = funcOp.getArgument(argIndex);
+
+        // Insert a new temporary argument with the new memref type.
+        funcOp.front().insertArgument(argIndex,
+                                      static_cast<Type>(newMemRefType));
+        auto newMemRef = funcOp.getArgument(argIndex);
+
+        // Replace all uses of the old memref.
+        if (failed(replaceAllMemRefUsesWith(oldMemRef, /*newMemRef=*/newMemRef,
+                                            /*extraIndices=*/{},
+                                            /*indexRemap=*/layoutMap,
+                                            /*extraOperands=*/{},
+                                            /*symbolOperands=*/{}))) {
+          // If it failed (due to escapes for example), bail out.
+          // Removing the temporary argument inserted previously.
+          funcOp.front().eraseArgument(argIndex);
+          continue;
+        }
+
+        // Since in this pass the objective is to normalize arguments of a
+        // function and replace the uses accordingly, we check if the function
+        // return type uses the same old memref type.
+        // TODO(avarmapml): Check - A function's return type might have a
+        // different memref layout and a map.
```

Comment:
It's not clear what this means - did you mean a non-identity layout map?

Reply:
So, the logic I've implemented checks if the return type has a memory layout that matches the old non-trivial memref layout in argument.
If there is a match an update (trivializing) is made to the return type.

But the return type might use a different memref layout (resulting from, say, alloc or a result of another function call) that should be normalized. I need to test these cases and accordingly handle them.